src/chain/split.rs

```diff
@@ -0,0 +1,939 @@
+        f.debug_struct(&format!("SplitOutputs<{}>", std::any::type_name::<T>()))
+            .field("scope", &self.scope)
+            .field("source", &self.source)
+            .field("last_key", &self.last_key)
+            .field("used", &self.used)
+            .finish()
+    }
+}
+
+impl<T: Splittable> SplitOutputs<T> {
+    /// Resume building connections for this split.
+    pub fn build<'w, 's, 'a, 'b>(
+        self,
+        builder: &'b mut Builder<'w, 's, 'a>,
+    ) -> SplitBuilder<'w, 's, 'a, 'b, T> {
+        assert_eq!(self.scope, builder.scope);
+        SplitBuilder {
+            outputs: self,
+            builder,
+        }
+    }
+
+    pub(crate) fn new(scope: Entity, source: Entity) -> Self {
+        Self {
+            scope,
+            source,
+            last_key: None,
+            used: Default::default(),
+        }
+    }
+}
+
+/// This is a type alias for the result returned by chainable [`SplitBuilder`]
+/// functions. If the last connection succeeded, you will receive [`Ok`] with
+/// the [`SplitBuilder`] which you can keep building off of. Otherwise if the
+/// last connection failed, you will receive an [`Err`] with the [`SplitBuilder`]
+/// bundled with [`SplitConnectionError`] to tell you what went wrong. You can
+/// continue building with the [`SplitBuilder`] even if an error occurred.
+///
+/// You can use `ignore_result` from the [`IgnoreSplitChainResult`] trait to
+/// just keep chaining without checking whether the connection suceeded.
+pub type SplitChainResult<'w, 's, 'a, 'b, T> = Result<
+    SplitBuilder<'w, 's, 'a, 'b, T>,
+    (SplitBuilder<'w, 's, 'a, 'b, T>, SplitConnectionError),
+>;
+
+/// A helper trait that allows users to ignore any failures while chaining
+/// connections to a split.
+pub trait IgnoreSplitChainResult<'w, 's, 'a, 'b, T: Splittable> {
+    /// Ignore whether the result of the chaining connection was [`Ok`] or [`Err`]
+    /// and just keep chaining.
+    fn ignore_result(self) -> SplitBuilder<'w, 's, 'a, 'b, T>;
+}
+
+impl<'w, 's, 'a, 'b, T: Splittable> IgnoreSplitChainResult<'w, 's, 'a, 'b, T>
+    for SplitChainResult<'w, 's, 'a, 'b, T>
+{
+    fn ignore_result(self) -> SplitBuilder<'w, 's, 'a, 'b, T> {
+        match self {
+            Ok(split) => split,
+            Err((split, _)) => split,
+        }
+    }
+}
+
+/// Information about why a connection to a split failed
+#[derive(ThisError, Debug, Clone)]
+#[error("An error occurred while trying to connect to a split")]
+pub enum SplitConnectionError {
+    /// The requested index was already connected to.
+    KeyAlreadyUsed,
+    /// The requested index is out of bounds.
+    KeyOutOfBounds,
+}
+
+/// Used by implementers of the [`Splittable`] trait to help them send their
+/// split values to the proper input slots.
+pub struct SplitDispatcher<'a, Key, Item> {
+    pub(crate) connections: &'a HashMap<Key, usize>,
+    pub(crate) outputs: &'a mut Vec<Vec<Item>>,
```

Comment:
I don't know the reasoning for this structure, at first glance, it seems that it should be possible to use `HashMap<Key, Vec<Item>>` instead of mapping to an index of a vec?

Reply:
There are a few reasons for this:

1. `HashMap` can have scatter the allocation of its buckets whereas `Vec` is guaranteed to use contiguous memory which reduces cache misses
2. We know the exact size of `ForkTargetStorage` in advance so we can pre-size this `Vec` to match its elements with `ForkTargetStorage`
3. Inside `OperateSplit::execute` we can iterate over this `Vec` zipped with `ForkTargetStorage` to drain the outputs and send them to their appropriate targets

Reply:
I'll also mention that it's very important for us to store the target entities in `ForkTargetStorage` to ensure that the connect operation and reachability calculations work correctly.

Reply:
> HashMap can have scatter the allocation of its buckets whereas Vec is guaranteed to use contiguous memory which reduces cache misses

I think this case is abit different, `outputs` is a vec of vecs, each of the inner `Vec<Item>` would have its own memory allocation.

Reply:
That's true, but `Vec<Vec<T>>` will still be less scattered than `HashMap<K, Vec<T>>`. I've considered using `Vec<SmallVec<[T; 1]>>` since most of the time I expect only one output per key, but that can be saved for future iterations on performance.

But the main reason to prefer `Vec` here is because `HashMap` would force us to do a second key lookup when it's time to drain the outputs.